plumbing/msg/sender.go

```diff
@@ -2,160 +2,25 @@ package msg
-		chainState: chainReader,
-		cst:        cst,
-		blockTimer: blockTimer,
-		inbox:      msgPool,
-		outbox:     msgQueue,
-		validator:  validator,
-		publish:    publish,
-	}
+// NewSender creates a new sender plumbing
+func NewSender(outbox *core.Outbox) *Sender {
+	return &Sender{outbox}
 }
 
 // Send sends a message. See api description.
 func (s *Sender) Send(ctx context.Context, from, to address.Address, value *types.AttoFIL, gasPrice types.AttoFIL, gasLimit types.GasUnits, method string, params ...interface{}) (out cid.Cid, err error) {
-	defer func() {
-		if err != nil {
-			msgSendErrCt.Inc(ctx, 1)
-		}
-	}()
-
-	encodedParams, err := abi.ToEncodedValues(params...)
-	if err != nil {
-		return cid.Undef, errors.Wrap(err, "invalid params")
-	}
-
-	// Lock to avoid race for message nonce.
-	s.l.Lock()
-	defer s.l.Unlock()
-
-	st, err := chain.LatestState(ctx, s.chainState, s.cst)
-	if err != nil {
-		return cid.Undef, errors.Wrap(err, "failed to load state from chain")
-	}
-
-	fromActor, err := st.GetActor(ctx, from)
-	if err != nil {
-		return cid.Undef, errors.Wrapf(err, "no actor at address %s", from)
-	}
-
-	nonce, err := nextNonce(fromActor, s.outbox, from)
-	if err != nil {
-		return cid.Undef, errors.Wrapf(err, "failed calculating nonce for actor %s", from)
-	}
-
-	msg := types.NewMessage(from, to, nonce, value, method, encodedParams)
-	smsg, err := types.NewSignedMessage(*msg, s.signer, gasPrice, gasLimit)
-	if err != nil {
-		return cid.Undef, errors.Wrap(err, "failed to sign message")
-	}
-
-	err = s.validator.Validate(ctx, smsg, fromActor)
-	if err != nil {
-		return cid.Undef, errors.Wrap(err, "invalid message")
-	}
-
-	smsgdata, err := smsg.Marshal()
-	if err != nil {
-		return cid.Undef, errors.Wrap(err, "failed to marshal message")
-	}
-
-	height, err := s.blockTimer.BlockHeight()
-	if err != nil {
-		return cid.Undef, errors.Wrap(err, "failed to get block height")
-	}
-
-	// Add to the local message queue/pool at the last possible moment before broadcasting to network.
-	if err := s.outbox.Enqueue(smsg, height); err != nil {
-		return cid.Undef, errors.Wrap(err, "failed to add message to outbound queue")
-	}
-	if _, err := s.inbox.Add(ctx, smsg); err != nil {
-		return cid.Undef, errors.Wrap(err, "failed to add message to message pool")
-	}
-
-	if err = s.publish(Topic, smsgdata); err != nil {
-		return cid.Undef, errors.Wrap(err, "failed to publish message to network")
-	}
-
-	log.Debugf("MessageSend with message: %s", smsg)
-	return smsg.Cid()
-}
-
-// nextNonce returns the next expected nonce value for an account actor. This is the larger
-// of the actor's nonce value, or one greater than the largest nonce from the actor found in the message pool.
-func nextNonce(act *actor.Actor, outbox *core.MessageQueue, address address.Address) (uint64, error) {
-	actorNonce, err := actor.NextNonce(act)
-	if err != nil {
-		return 0, err
-	}
+	return s.outbox.Send(ctx, from, to, value, gasPrice, gasLimit, method, params...)
```

Comment:
Let's just ditch this file completely and put the outbox on the plumbing API struct directly like the wallet.

Reply:
Done 🎉